plugins/dell-dock2/fu-dell-dock2-wtpd.c

```diff
@@ -0,0 +1,181 @@
+
+	/* IDs */
+	instance_id = g_strdup_printf("WTPD\\DOCKTYPE_%02x&DOCKSKU_%02x", dock_type, dock_sku);
+	fu_device_add_instance_id(device, instance_id);
+	fu_device_set_logical_id(device, devname);
+
+	/* version */
+	wtpd_version = fu_dell_dock2_ec_get_wtpd_version(proxy);
+	dynamic_version = fu_dell_dock2_wtpd_ver_string(wtpd_version);
+	fu_device_set_version(device, dynamic_version);
+
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock2_wtpd_write(FuDevice *device,
+			 FuFirmware *firmware,
+			 FuProgress *progress,
+			 FwupdInstallFlags flags,
+			 GError **error)
+{
+	FuDevice *proxy = fu_device_get_proxy(device);
+	g_autoptr(GBytes) fw = NULL;
+	g_autoptr(GBytes) fw_whdr = NULL;
+	g_autoptr(FuChunkArray) chunks = NULL;
+	g_autofree gchar *dynamic_version = NULL;
+	g_autoptr(GByteArray) res = g_byte_array_sized_new(DELL_DOCK2_EC_HID_RESPONSE_LENGTH);
+
+	/* basic tests */
+	g_return_val_if_fail(device != NULL, FALSE);
+	g_return_val_if_fail(FU_IS_FIRMWARE(firmware), FALSE);
+
+	/* get default firmware image */
+	fw = fu_firmware_get_bytes(firmware, error);
+	if (fw == NULL)
+		return FALSE;
+
+	/* construct writing buffer */
+	fw_whdr = fu_dell_dock2_hid_fwup_pkg_new(fw, DELL_DOCK2_EC_DEV_TYPE_WTPD, 0);
+
+	/* prepare the chunks */
+	chunks = fu_chunk_array_new_from_bytes(fw_whdr, 0, DELL_DOCK2_EC_HID_DATA_PAGE_SZ);
+
+	/* write to device */
+	for (guint i = 0; i < fu_chunk_array_length(chunks); i++) {
+		g_autoptr(FuChunk) chk = NULL;
+
+		chk = fu_chunk_array_index(chunks, i, error);
+		if (chk == NULL)
+			return FALSE;
+
+		if (!fu_dell_dock2_ec_hid_write(proxy, fu_chunk_get_bytes(chk), error))
+			return FALSE;
+	}
+
+	/* request parent EC to reset the dock */
+	fu_device_add_flag(proxy, FWUPD_DEVICE_FLAG_NEEDS_ACTIVATION);
+
+	/* dock will reboot to re-read; this is to appease the daemon */
+	g_debug("pd firmware written successfully; waiting for dock to reboot");
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock2_wtpd_open(FuDevice *device, GError **error)
+{
+	return fu_device_open(fu_device_get_proxy(device), error);
+}
+
+static gboolean
+fu_dell_dock2_wtpd_close(FuDevice *device, GError **error)
+{
+	return fu_device_close(fu_device_get_proxy(device), error);
+}
+
+static void
+fu_dell_dock2_wtpd_finalize(GObject *object)
+{
+	G_OBJECT_CLASS(fu_dell_dock2_wtpd_parent_class)->finalize(object);
+}
+
+static void
+fu_dell_dock2_wtpd_set_progress(FuDevice *self, FuProgress *progress)
+{
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 13, "detach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 72, "write");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 9, "attach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 7, "reload");
+}
+
+static void
+fu_dell_dock2_wtpd_init(FuDellDock2Wtpd *self)
+{
+	fu_device_add_protocol(FU_DEVICE(self), "com.dell.dock2");
+	fu_device_set_vendor(FU_DEVICE(self), "Dell Inc.");
```

Comment:
do you really need the "Inc."? Do you need it at all if you add a quirk-only GUID of `USB//VID_413C` in probe or setup then fwupd will do the correct thing.

Reply:
"Dell Inc." used to live in `dell-dock.quirk` file so conitnued here, maybe because the dmi id?
```
$ cat /sys/class/dmi/id/sys_vendor 
Dell Inc.
```

you mean below lines eventually do the same thing?
```
fu_device_set_vendor(FU_DEVICE(self), "Dell Inc.");
fu_device_add_vendor_id(device, "USB:0x413C");
```